osmflatc/src/strings.rs

```diff
@@ -1,9 +1,63 @@
 use ahash::AHashMap;
-use inlinable_string::{InlinableString, StringExt};
+
+#[derive(Debug, Clone, Copy)]
+struct TerminatedStringPtr {
+    ptr: *const u8,
+}
+
+// We use this (unsafe) wrapper to get the most compact hashmap possible
+// Using a "&'static str" would be bigger due to the length stored
+// Using a String would allocate a lot of individual blocks
+// Using a small-string-optimized structure would create large objects
+impl TerminatedStringPtr {
+    /// Safety:
+    /// Requires the data pointed to to:
+    /// * Be \0 terminated
+    /// * Outlive TerminatedStringPtr
+    unsafe fn from_ptr(ptr: *const u8) -> Self {
+        Self { ptr }
+    }
+
+    fn as_bytes(&self) -> &[u8] {
+        // Safety:
+        // If constructed properly from a 0-terminated string that outlives this instance this is safe
+        unsafe { std::ffi::CStr::from_ptr(self.ptr as *const i8).to_bytes() }
+    }
+}
+
+impl PartialEq for TerminatedStringPtr {
+    fn eq(&self, other: &TerminatedStringPtr) -> bool {
+        self.as_bytes() == other.as_bytes()
+    }
+}
+
+impl std::hash::Hash for TerminatedStringPtr {
+    fn hash<H>(&self, h: &mut H)
+    where
+        H: std::hash::Hasher,
+    {
+        self.as_bytes().hash(h)
+    }
+}
+
+impl Eq for TerminatedStringPtr {}
+
+impl std::borrow::Borrow<[u8]> for TerminatedStringPtr {
+    fn borrow(&self) -> &[u8] {
+        self.as_bytes()
+    }
+}
 
 #[derive(Debug, Default)]
 pub struct StringTable {
-    indexed_data: AHashMap<InlinableString, u64>,
+    // Append only, we will never reallocate any data inside
+    data: Vec<Vec<u8>>,
```

Comment:
You could store Box<[u8]> in the outer `Vec`, which tells you that you can't change it anymore. It is also smaller than Vec by one word (no capacity).

Reply:
Yes, but that saves only 8 bytes per 4 MB.